### PR TITLE
Add settings similarity search

### PR DESCRIPTION
**Pull request by rzhao271**

### CODE REVIEW

**Comment by TylerLeonhardt** [comment]
Usage looks fine from my perspective but since the code is all in the preferences code, I'll defer to @roblourens for approval.

File: src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts [hunk 1]
```diff
@@ -305,4 +297,104 @@ export class SettingMatches {
...
+			return Promise.resolve(null);
+		}
+
+		this._keysProvider.updateModel(preferencesModel);
+		const settingKeys = this._keysProvider.getSettingKeys();
+		const settingsRecord = this._keysProvider.getSettingsRecord();
+
+		return this.semanticSimilarityService.getSimilarityScore(this._filter, settingKeys, token ?? CancellationToken.None).then((scores) => {
+			const filterMatches: ISettingMatch[] = [];
+			const foundKeys = new Set<string>();
```
Comment (rzhao271) [inline src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts:372]: Something for me to confirm is whether the current list of keys contains duplicates.

File: src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts [hunk 2]
```diff
@@ -305,4 +297,103 @@ export class SettingMatches {
...
+	searchModel(preferencesModel: ISettingsEditorModel, token?: CancellationToken | undefined): Promise<ISearchResult | null> {
+		if (!this.semanticSimilarityService.isEnabled() || !this._filter) {
+			return Promise.resolve(null);
+		}
+
+		this._keysProvider.updateModel(preferencesModel);
+		const settingKeys = this._keysProvider.getSettingKeys();
+		const settingsRecord = this._keysProvider.getSettingsRecord();
+
+		return this.semanticSimilarityService.getSimilarityScore(this._filter, settingKeys, token ?? CancellationToken.None).then((scores) => {
```
Comment (roblourens) [inline src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts:373]: async/await?

File: src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts [hunk 3]
```diff
@@ -305,4 +297,103 @@ export class SettingMatches {
...
+	updateModel(preferencesModel: ISettingsEditorModel) {
+		if (preferencesModel === this.currentPreferencesModel) {
+			return;
+		}
+
+		this.currentPreferencesModel = preferencesModel;
+		this.refresh();
+	}
+
+	refresh() {
```
Comment (roblourens) [inline src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts:317]: private?

File: src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts [hunk 4]
```diff
@@ -305,4 +297,103 @@ export class SettingMatches {
 	}
 }
 
+class RemoteSearchProviderKeyCache {
```
Comment (roblourens) [inline src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts:300]: This doesn't really seem like a "cache", it just presents the same data from the ISettingsEditorModel in a different shape. Should this just be available from the settings editor model directly somehow?

**Review by roblourens — DISMISSED** [review]

File: src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts (hunk 4 above)
Comment (rzhao271) [reply src/vs/workbench/contrib/preferences/browser/preferencesSearch.ts:300]: Considering it's only used for search, I'll leave the implementation here for now and will rename the class.

**Review by roblourens — APPROVED** [review]

**Pull request merged**